Update/jetpack cloud backup status with new date management

cleacos opened the pull request:
#### Changes proposed in this Pull Request

Add a date index to the backup log to facilitate the manipulations of the backups. The backups will work in the site time zone like Activity Log works in Calypso.

When we process all the backups (one time), we'll get the oldest backup available, so we can limit the DatePicker on the left navigation.

**This PR is based on https://github.com/Automattic/wp-calypso/pull/40149 Still in progress...**

#### Testing instructions

- TBD

cleacos committed:
Improving the DatePicker component to match the specification

- Keeping the component neutral: managing directly with the Date objects
- Show the appropriate date format (Today, Yesterday, with year or not)
- Applied the expected date selection behavior: < previous [current date] next >

cleacos committed:
Change the selectedDate to a Date object instead of manage strings

- Updated all the logic to use Date object
- Improved the connection with the store, determining if we have a siteId
selected, if not, redirect to parent URL, /backups
- Commented temporally some code to allow the review for this PR and to
pass the Linter without errors.

cleacos committed:
Add a backup logs index to facilitate the manipulations of the backups

- Created a indexed list by date of backups/activity
- The date of the backups is moved to the site time zone of offset
- Added date limit on the DatePicker left navigation (based on the last backup date available)

cleacos committed:
Reset the Backup Status state when we switch the site

- Other minor improvements

enejb commented on client/landing/jetpack-cloud/components/date-picker/index.jsx:
```diff
@@ -18,80 +19,107 @@ import Gridicon from 'components/gridicon';
 			<div className="date-picker">
 				<Button compact borderless onClick={ this.shuttleLeft }>
-					<Gridicon icon="chevron-left" />
+					<Gridicon icon="chevron-left" className={ ! this.canShuttleLeft() && 'disabled' } />
```
Why even show this at all? instead of disabeling it?

cleacos replied:
That is like the design. I prefer the option of hiding it. We could discuss it with the design team.

cleacos committed:
Refactor method name shuttle[Right|Left] to goTo[Next|Previous]Date

cleacos committed:
Fix unmounted BackupStatus should refresh the state when it's mounted

cbauerman commented on client/landing/jetpack-cloud/sections/backups/main.jsx:
```diff
@@ -3,95 +3,227 @@
+	/**
+	 * Create an indexed log of backups based on the date of the backup and in the site time zone
+	 */
+	createIndexedLog() {
```
I like the idea of creating an index with the dates! However, the way they are [placed into state reminds me of a lot of `getDerivedState`, and the article You Probably Don't Need Derived State](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#when-to-use-derived-state).

I feel like this method is here because of the access to the localized moment, but it actual has very little to do with the state of the component ( for example changing the selected date does not change anything about the index ). I would think this fits in much better in the `connect` function since it is basically another property ( retrieve the logs and transform them into the index in the connect function, then only pass the index and oldest date as properties )

cleacos replied:
I'm going to move the method outside of the class and those fields on the state to the props

cleacos committed:
Move the method createIndexedLog from the class and call it from connect

 - indexedLog and oldestDateAvailable will be a prop instead of a field in the state

cleacos committed:
Remove the commented code to facilitate the review

cleacos commented:
Close this PR in favor of https://github.com/Automattic/wp-calypso/pull/40403
